### DISCUSSION TITLE
Best practice for persisting state

### DISCUSSION THREAD

**Discussion by stefanfrede** [statelyai/xstate#1902] [Q&A]
I started to try to wrap my head around XState.

With the help of some screencasts and going through the docs I managed to setup a [fetch machine](https://github.com/stefanfrede/the-fab-5/blob/main/src/machines/fetch.js).

In my [playground](https://github.com/stefanfrede/the-fab-5) I grab some data in one component and based on that data I grab additional data in another component.

Now I would like to persist my data so that once I have it I don't fetch it again. And with that I somehow struggle.

I tried to [persist and rehydrate the state](https://xstate.js.org/docs/packages/xstate-vue/#persisted-and-rehydrated-state) but that was only working in one componet (in the component where I fetch the cover art I got errors).

At the moment my idea is to set up some kind of additional store (vuex or just a plain old object) where I add my results and check before fetching if my  results are existing.

Is this a good idea? Or to return back to my subject, what would be considered the best practice to persist my state with XState?

Any push in the right direction is much appreciated!

~Because I find it quite difficult to figure out how to best setup XState in a project it would be awesome if someone could additionally point me somewhere where I can figure this out.~ Should have taken a look at the [resources](https://xstate.js.org/docs/about/resources.html) before.

Thanks a ton in advance!

**Comment by Andarist** [comment]
If you want to implement caching strategies and when you have short-lived machines then you should externalize the cache - I would probably handle it in your `fetchData` actor which could be implemented using machines or not (it won't matter for your `fetchMachine`)

**Comment by tomByrer** [comment]
I think there are confusing & conflicting therms with XState;
"state" = the position of your XState machine?  Or your (fetched) data?

**Reply by stefanfrede** [reply]
I guess I mixed it up somehow.

As far as I understand it correctly if I persist the state (machine) then my context (fetched data) will get persisted too.

So, after rehydrating my machine I also have the data I already fetched.

This works fine in my example for fetching the records data but not for fetching the cover art (besides throwing an error I additionally don't want to persist a dozen machines).

At the moment I'm reading up on invoking services and actors because my idea is to have a parent machine which I feed my fetched data from my child machines (hopefully this is a good idea).

**Reply by Andarist** [reply]
> As far as I understand it correctly if I persist the state (machine) then my context (fetched data) will get persisted too.

Correct.

---

If you plan to persist a state of more than a root machine but also states of all of your children machines etc then this is not possible easily right now **but** there are plans for allowing this and we've been talking about possible APIs, semantics etc just this week. So if you are interested in that - stay tuned!
